### PR TITLE
VMR E2E tests for applying patches

### PR DESCRIPTION
**Pull request by MilenaHristova**
https://github.com/dotnet/arcade/issues/10798

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        @"{{
+  ""patchesPath"": ""{0}"",
+  ""additionalMappings"": [
+    {1}
+  ],
+
+  ""defaults"": {{
+      ""defaultRef"": ""main"",
+      ""exclude"": [
+        ""**/*.dll"",
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs:45]: Can you add a test / extend existing and put a .dll file in a repo and verify it's not synchronized into the VMR? I think that would be very valuable.

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        ""**/*.nupkg""
+      ]
+    }},
+
+  ""mappings"": [
+    {2}
+  ]
+}}";
+
+    public static readonly string MappingTemplate =
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs:60]: Did you consider having the `source-mappings.json` as C# objects and only serialize them into JSON? Might avoid this JSON kung-fu this way?

We even have classes for that:
https://github.com/dotnet/arcade-services/blob/4b97ba07fa0b7cbda645aa94f32099c4e3ef0710/src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/SourceMappingParser.cs#L155-L170

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/GitOperations.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.DotNet.DarcLib.Helpers;
+using Microsoft.Extensions.Logging.Abstractions;
+
+#nullable enable
+namespace Microsoft.DotNet.Darc.Tests.VirtualMonoRepo;
+
+public class GitOperations
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/GitOperations.cs:16]:
```suggestion
internal class GitOperations
```

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Models/Dependency.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.DotNet.DarcLib.Helpers;
+
+namespace Microsoft.DotNet.Darc.Tests.VirtualMonoRepo;
+
+public record Dependency(string Name, LocalPath Uri);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Models/Dependency.cs:9]:
```suggestion
internal record Dependency(string Name, LocalPath Uri);
```

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncRepoChangesE2ETest.cs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        await InitializeSubmodule(_privateRepoPath, "submodule1", _externalRepoPath, submoduleRelativePath);
+        await CommitAll(_privateRepoPath, "Add submodule");
+        await UpdateRepoToLastCommit("test-repo", _privateRepoPath);
+
+        var expectedFilesFromRepos = new List<LocalPath>
+        {
+            testRepoFilePath,
+            dependencyFilePath,
+            submoduleFilePath,
+            _vmrPath / "src" / "test-repo" / ".gitmodules",
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncRepoChangesE2ETest.cs:101]: I wouldn't test for this file. We don't care about it and in future we might even remove these and we don't care
```suggestion
```

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        return await GetRepoLastCommit(repoPath);
+    }
+
+    public static string EscapePath(string path)
+    {
+        return path.Replace("\\", "\\\\");
+    }
+
+    public record SourceMapping(string Name, string DefaultRemote, List<string>? Exclude = null);
+    public record AdditionalMapping(string Source, string Destination);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs:273]:
```suggestion
    internal record SourceMapping(string Name, string DefaultRemote, List<string>? Exclude = null);
    internal record AdditionalMapping(string Source, string Destination);
```

Comment (premun) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs:273]: Also, should these go into the `Models` maybe?

Comment (premun) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs:273]: Maybe just one common file `Models.cs` would be enough for everything?

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrPatchesE2ETestsBase.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    protected override async Task CopyVmrForCurrentTest()
+    {
+        CopyDirectory(VmrE2ETestsOneTimeSetUp.commonVmrPath, _vmrPath);
+
+        var mappings = new List<SourceMapping>
+        {
+            new SourceMapping(installerRepoName, _installerRepoPath.Path.Replace("\\", "\\\\")),
+            new SourceMapping(testRepoName, _privateRepoPath.Path.Replace("\\", "\\\\"))
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrPatchesE2ETestsBase.cs:32]: Why do we need these escapes?

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrPatchesE2ETestsBase.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        };
+        var sm = GenerateSourceMappings(mappings, "src/installer/patches/");
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrPatchesE2ETestsBase.cs:34]: Is `installer` one of the repositories we're synchronizing in? (it should be)

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncAdditionalMappingsE2ETest.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.DotNet.DarcLib.Helpers;
+using NUnit.Framework;
+
+#nullable enable
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncAdditionalMappingsE2ETest.cs:11]: You can remove these `#nullable enable` statements as the whole project has nullability turned on

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+            }
+        }
+
+        var versionDetails = string.Format(Constants.VersionDetailsTemplate, dependenciesString);
+        File.WriteAllText(repoPath / "eng" / _versionDetailsName, versionDetails);
+        await CommitAll(repoPath, "update version details");
+        return await GetRepoLastCommit(repoPath);
+    }
+
+    public static string EscapePath(string path)
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs:267]: I'd turn this into an extension method and then you can do:
```
var path = "....";
path = path.Escape();
```

Also I see this escaping all over the place so you please go through those and use this method.

I am also curious why we need to keep escaping these?

**Review by premun — COMMENTED** [review]
Good job! We're nearly there!

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs (hunk 2 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs:60]: they're private though. I can make them public and move them to the Models?

Comment (premun) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs:60]: Yes please

**Comment by premun** [comment]
@MilenaHristova I am writing a test for my bug to test this and I think have an opinion about the contents of the file being in C# vs a file.

It would be nice if the content was in a file :)

And we could get/compare the content easily in a test, e.g. something like `CompareFileContents(string path, string resourceFileName)` and `GetResourceFileContent(string resourceFileName)`.

**Comment by MilenaHristova** [comment]
> @MilenaHristova I am writing a test for my bug to test this and I think have an opinion about the contents of the file being in C# vs a file.
> 
> It would be nice if the content was in a file :)
> 
> And we could get/compare the content easily in a test, e.g. something like `CompareFileContents(string path, string resourceFileName)` and `GetResourceFileContent(string resourceFileName)`.
> 
> I will be adding tests that would duplicate this a lot and also the content is quite tied to the content in the patch files so it seems weird to have the split there.

ok i'll move it to a file

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+using System.Threading.Tasks;
+using CSharpx;
+using FluentAssertions;
+using Microsoft.DotNet.DarcLib.Helpers;
+using NUnit.Framework;
+
+#nullable enable
+namespace Microsoft.DotNet.Darc.Tests.VirtualMonoRepo;
+
+public abstract class VmrE2ETestsBase : GitOperations
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs:20]: Instead of inheriting from `GitOperations`, it's better to create a class like `GitOperationsHelper` and then create a field in the `VmrE2ETestsBase` class:

```csharp
protected GitOperationsHelper GitOperations { get; } = new();
```

then you can call it from the tests:

```csharp
GitOperations.InitialCommit(/* ... */)
```

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using CSharpx;
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs:12]: What's `CSharpx` and why do we need it?

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsOneTimeSetUp.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+public class VmrE2ETestsOneTimeSetUp : GitOperations
+{
+    public static readonly LocalPath testsDirectory;
+    public static readonly LocalPath commonVmrPath;
+    public static readonly LocalPath commonPrivateRepoPath;
+    public static readonly LocalPath commonSubmodulePath;
+    public static readonly LocalPath commonDependencyPath;
+    public static readonly LocalPath commonInstallerPath;
+    public static readonly LocalPath commonExternalRepoPath;
+    public static readonly LocalPath repoWithSubmodulePath;
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsOneTimeSetUp.cs:25]: These should be capitalized

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs (hunk 12 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrE2ETestsBase.cs]: no idea, I guess we don't

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrPatchChangingFileTest.cs [hunk 14]
```diff
@@ -45,28 +45,18 @@ public async Task PatchesAreAppliedTest()
...
-AAA
-CCC
-end of changes");
+        CompareFileContents(testRepoFilePath, "test-file-after-patch.txt");
         await CheckAllIsCommited(_vmrPath);
 
         // a change in the patch
 
-        //File.Copy(_resourcesPath / "changed-patch.patch", _installerRepoPath / "patches" / "test-repo" / "example.patch", true);
+        //File.Copy(VmrTestsOneTimeSetUp.ResourcesPath / "changed-patch.patch", _installerRepoPath / "patches" / "test-repo" / "example.patch", true);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrPatchChangingFileTest.cs:53]: I guess don't forget about this

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncAdditionalMappingsTest.cs [hunk 15]
```diff
@@ -62,14 +62,14 @@ protected override async Task CopyReposForCurrentTest()
         File.WriteAllText(
             _filePath,
             "A file that needs to be copied outside of the src folder");
-        File.WriteAllText(_specialRepoPath / "eng" / _versionDetailsName, Constants.EmptyVersionDetails);
+        File.WriteAllText(_specialRepoPath / "eng" / Constants.VersionDetailsName, Constants.EmptyVersionDetails);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncAdditionalMappingsTest.cs:65]: There's a class called `VersionFiles` and this should come from there.

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs [hunk 16]
```diff
@@ -71,4 +71,9 @@ public class Constants
     ""source"": ""{0}"",
     ""destination"": ""{1}""
 }}";
+
+    public static readonly string VersionDetailsName = "Version.Details.xml";
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs:75]: Same, should be from the `VersionFiles` class

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsBase.cs [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+        {
+            vmrPath / "git-info" / "AllRepoVersions.props",
+            vmrPath / "src" / "source-manifest.json",
+            vmrPath / "src" / "source-mappings.json",
+        };
+
+        foreach (var repo in syncedRepos)
+        {
+            expectedFiles.Add(vmrPath / "src" / repo / "eng" / Constants.VersionDetailsName);
+            expectedFiles.Add(vmrPath / "git-info" / $"{repo}.props");
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsBase.cs:88]: All `git-info` names should come from `VmrInfo` where we have the name of this folder (it will actually change soon so that's why we care)

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsBase.cs [hunk 18]
```diff
@@ -7,16 +7,24 @@
...
+using Microsoft.Extensions.Logging;
 using NUnit.Framework;
 
 #nullable enable
 namespace Microsoft.DotNet.Darc.Tests.VirtualMonoRepo;
 
 public abstract class VmrTestsBase
 {
+    private Lazy<IServiceProvider> _serviceProvider = null!;
+    private readonly CancellationTokenSource _cancellationToken = new();
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsBase.cs:27]: Instead of this, I'd just pass `default` to the Vmr classes which equals to `CancellationToken.None`

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsBase.cs [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+    protected abstract Task CopyReposForCurrentTest();
+
+    protected abstract Task CopyVmrForCurrentTest();
+
+    private IServiceProvider CreateServiceProvider() => new ServiceCollection()
+        .AddTransient<GitFileManagerFactory>()
+        .AddLogging(b => b.AddConsole().AddFilter(l => l >= LogLevel.Information))
+        .AddVmrManagers(
+        sp => sp.GetRequiredService<GitFileManagerFactory>(),
+        "git",
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsBase.cs:85]:
```suggestion
        _processManager.GitExecutable,
```

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsOneTimeSetUp.cs [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    public static readonly LocalPath CommonExternalRepoPath;
+    public static readonly LocalPath RepoWithSubmodulePath;
+    public static readonly LocalPath ResourcesPath;
+    private GitOperationsHelper _gitOperations { get; } = new();
+
+
+    static VmrTestsOneTimeSetUp()
+    {
+        var assembly = Assembly.GetAssembly(typeof(VmrTestsBase)) ?? throw new Exception("Assembly not found");
+        ResourcesPath = new NativePath(Path.Join(Path.GetDirectoryName(assembly.Location), "Resources"));
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTestsOneTimeSetUp.cs:35]:
```suggestion
        ResourcesPath = new NativePath(Path.GetDirectoryName(assembly.Location)) / "Resources");
```

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncAdditionalMappingsTest.cs (hunk 15 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrSyncAdditionalMappingsTest.cs:65]: Check this commit - https://github.com/dotnet/arcade-services/pull/2154/commits/6cc30bb8610788c0bed30a8e96188bfc2c3570d6
Something similar should this be done in this PR

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+@"<?xml version=""1.0"" encoding=""utf-8""?>
+<Dependencies>
+    <ProductDependencies>
+        {0}
+    </ProductDependencies>
+    <ToolsetDependencies>
+    </ToolsetDependencies>
+</Dependencies>";
+
+    public static readonly string EmptyVersionDetails =
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs:19]: I think this is still unresolved

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/GitOperationsHelper.cs [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    public async Task ChangeSubmoduleUrl(LocalPath repo, LocalPath submodulePath, LocalPath newUrl)
+    {
+        await ProcessManager.ExecuteGit(repo, "submodule", "set-url", submodulePath, newUrl);
+    }
+
+    private async Task ConfigureGit(LocalPath repo)
+    {
+        await ProcessManager.ExecuteGit(repo, "config", "user.email", DarcLib.Constants.DarcBotEmail);
+        await ProcessManager.ExecuteGit(repo, "config", "user.name", DarcLib.Constants.DarcBotName);
+    }
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/GitOperationsHelper.cs:98]: Btw this is also an option:

```suggestion
    public Task RemoveSubmodule(LocalPath repo, string submoduleRelativePath)
    {
        return ProcessManager.ExecuteGit(repo, "rm", "-f", submoduleRelativePath);
    }

    public Task PullMain(LocalPath repo)
    {
        return ProcessManager.ExecuteGit(repo, "pull", "origin", "main");
    }

    public Task ChangeSubmoduleUrl(LocalPath repo, LocalPath submodulePath, LocalPath newUrl)
    {
        return ProcessManager.ExecuteGit(repo, "submodule", "set-url", submodulePath, newUrl);
    }

    private async Task ConfigureGit(LocalPath repo)
    {
        await ProcessManager.ExecuteGit(repo, "config", "user.email", DarcLib.Constants.DarcBotEmail);
        await ProcessManager.ExecuteGit(repo, "config", "user.name", DarcLib.Constants.DarcBotName);
    }
```

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Constants.cs:19]: O yes, I stashed that change

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Dependency.cs [hunk 23]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.DotNet.DarcLib.Helpers;
+
+namespace Microsoft.DotNet.Darc.Tests.VirtualMonoRepo;
+
+public record Dependency(string Name, LocalPath Uri);
```
[the remark this answers is not in the document]
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Dependency.cs:9]: but then the methods that use it should also be internal

[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/Dependency.cs:9]: Right. So, I guess everything but Test classes themselves should be ideally internal.

**Review by premun — APPROVED** [review]
